packages/botonic-api/src/rest/routes/validation/events.ts

```diff
@@ -0,0 +1,261 @@
+
+const inQuery: ParamSchema = { in: ['query'] }
+const inParams: ParamSchema = { in: ['params'] }
+const inBody: ParamSchema = { in: ['body'] }
+const isRequired: ValidatorsSchema = { notEmpty: true }
+const isOptional: ParamSchema = { optional: { options: { nullable: true } } }
+const isNaturalNumber: ParamSchema = {
+  isNumeric: { options: { no_symbols: true } },
+}
+const isNumeric: ParamSchema = { isNumeric: true }
+const isBoolean: ParamSchema = { isBoolean: true }
+const isDate: ParamSchema = { isDate: true }
+const isIn = (valueList: string[]): ParamSchema => {
+  return {
+    isIn: { options: [valueList] },
+    errorMessage: `Invalid value. Allowed values: ${valueList}`,
+  }
+}
+const equals = (text: string): ParamSchema => {
+  return {
+    equals: { options: [text] },
+    errorMessage: `Invalid value. Value should be '${text}'`,
+  }
+}
+const toInt: ParamSchema = { toInt: true }
+
+export const limitParamSchema: ParamSchema = {
+  ...inQuery,
+  ...isOptional,
+  ...isNaturalNumber,
+  ...toInt,
+}
+export const offsetParamSchema = limitParamSchema
+
+export const eventIdParamSchema: ParamSchema = { ...inParams, ...isRequired }
+
+export const buttonsSchema: Schema = {
+  buttons: {
+    ...inBody,
+    ...isOptional,
+    isArray: true,
+    errorMessage: 'buttons have to be an array',
+  },
+  'buttons.*.title': { ...inBody, ...isRequired },
+  'buttons.*.webview': { ...inBody, ...isOptional },
+  'buttons.*.params': { ...inBody, ...isOptional },
+  'buttons.*.url': { ...inBody, ...isOptional },
+  'buttons.*.target': { ...inBody, ...isOptional },
+  'buttons.*.payload': { ...inBody, ...isOptional },
+}
+
+export const repliesSchema: Schema = {
+  replies: {
+    ...inBody,
+    ...isOptional,
+    isArray: true,
+    errorMessage: 'replies have to be an array',
+  },
+  'replies.*.title': { ...inBody, ...isRequired },
+  'replies.*.payload': { ...inBody, ...isRequired },
+}
+
+export const carouselSchema: Schema = {
+  elements: {
+    ...inBody,
+    isArray: true,
+    errorMessage: 'elements have to be an array',
+  },
+  'elements.*.pic': { ...inBody, ...isRequired },
+  'elements.*.title': { ...inBody, ...isRequired },
+  'elements.*.subtitle': { ...inBody, ...isOptional },
+}
+
+export const baseEventSchema: Schema = {
+  eventId: { ...inBody, ...isRequired },
+  eventType: { ...inBody, ...isRequired, ...isIn(Object.values(EventTypes)) },
+  createdAt: { ...inBody, ...isRequired, ...isDate },
+  modifiedAt: { ...inBody, ...isOptional, ...isDate },
+}
+
+export const botonicMessageEventSchema: Schema = {
+  ...baseEventSchema,
+  ack: { ...inBody, ...isRequired, ...isIn(Object.values(MessageEventAck)) },
+  from: { ...inBody, ...isRequired, ...isIn(Object.values(MessageEventFrom)) },
+  type: { ...inBody, ...isRequired, ...isIn(Object.values(MessageEventTypes)) },
+  typing: { ...inBody, ...isRequired, ...isNaturalNumber, ...toInt },
+  delay: { ...inBody, ...isRequired, ...isNaturalNumber, ...toInt },
+}
```

Comment:
Yes, I thought the same, we'll be having duplicated code by having all the type definitions and this validations, but I haven't seen any easy/automatic way of taking advantage of the type definitions to generate the validations.

I've tried (not much) this approach: [https://github.com/ForbesLindesay/typescript-json-validator](https://github.com/ForbesLindesay/typescript-json-validator) but it seems it's not actively maintained and I couldn't generate the file with the validation, it crashed.

I've seen this other approach: [https://github.com/kimamula/ts-transformer-keys](https://github.com/kimamula/ts-transformer-keys), but I think it's too complicated and I don't think it will solve our problem by having only the keys of an interface...

Another approach would be to have all events/users as classes and add a validation method to it and execute this method when creating/updating one, but I think this way will change too much things right now.

Reply:
ok, let's leave it like this and we'll research alternatives in the future